src/bots/filler.ts

```diff
@@ -897,6 +900,72 @@ export class FillerBot implements Bot {
 		return true;
 	}
 
+	protected selectSubsetOfMakers(nodeToFill: NodeToFill) {
+		let makerNodes = nodeToFill.makerNodes;
+		if (makerNodes.length >= MAX_MAKERS_PER_FILL) {
+			// Calculate the values for each item and sum them up
+			const values: BN[] = [];
+			const totalValue = ZERO;
+			for (const node of makerNodes) {
+				if (node.order) {
+					const value = node.order.baseAssetAmount.sub(
+						node.order.baseAssetAmountFilled
+					);
+					values.push(value);
+					totalValue.add(value);
+				}
+			}
+
+			// Calculate probabilities of skipping for each item
+			const probabilities: number[] = values.map((value) =>
+				value.mul(PERCENTAGE_PRECISION).div(totalValue).toNumber()
+			);
+
+			// Select MAX_MAKERS_PER_FILL unique items, biasing towards the ones in the front
+			const selectedItems: DLOBNode[] = [];
+			let tries = 0;
+
+			while (selectedItems.length < MAX_MAKERS_PER_FILL && tries < 10) {
```

Comment:
i wonder if it's possible to do this O(n) instead of O(2n)

Reply:
can generate an array of choices based on the weighted probabilities and do a random index in the array, then don't also have the weird stop condition of `tries < 10`.

candlebot did something similar (but it's O(n^2) for how many possible choices there are, i think it can be optimized though

https://github.com/drift-labs/internal-keeper-bot/blob/master/src/bots/candlePainter.ts#L472-L481